Java/parkservices/src/main/java/com/amazon/randomcutforest/extendedrandomcutforest/threshold/ThresholdedRandomCutForest.java

```diff
@@ -0,0 +1,406 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.extendedrandomcutforest.threshold;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+
+import java.util.Arrays;
+
+import com.amazon.randomcutforest.RandomCutForest;
+import com.amazon.randomcutforest.extendedrandomcutforest.AnomalyDescriptor;
+import com.amazon.randomcutforest.returntypes.DiVector;
+
+public class ThresholdedRandomCutForest{
+
+    public static int MINIMUM_OBSERVATIONS_FOR_EXPECTED = 100;
+
+    // a parameter that determines if the current potential anomaly is describing the same anomaly
+    // within the same shingle or across different time points
+    protected double ignoreSimilarFactor = 0.3;
+    // uses attribution by default; can be useful without attribution as a general featurization
+    protected double triggerFactor = 3.0;
+
+    // saved attribution of the last seen anomaly
+    protected long lastAnomalyTimeStamp;
+    protected double lastAnomalyScore;
+    protected DiVector lastAnomalyAttribution;
+    protected double lastScore;
+    double[] lastAnomalyPoint;
+    double[] lastExpectedPoint;
+
+    boolean previousIsPotentialAnomaly;
+    boolean inHighScoreRegion;
+
+    // flag that determines if we should dedup similar anomalies not in the same shingle, for example an
+    // anomaly, with the same pattern is repeated across more than a shingle
+    protected boolean ignoreSimilar;
+
+    // for anomaly description we would only look at these may top attributors
+    // note that expected value is not well defined when this number is greater than 1
+    int numberOfAttributors = 2;
+
+    protected RandomCutForest forest;
+    protected BasicThresholder thresholder;
+
+
+
+    public ThresholdedRandomCutForest(RandomCutForest.Builder builder, double anomalyRate){
+        forest = builder.build();
+        checkArgument(!forest.isInternalShinglingEnabled(),"Incorrect setting, not supported");
+        thresholder = new BasicThresholder(anomalyRate);
+        if (forest.getDimensions()/forest.getShingleSize() == 1){
+            thresholder.setLowerThreshold(1.1);
+        }
+    }
+
+    public ThresholdedRandomCutForest(RandomCutForest forest, BasicThresholder thresholder, ThresholdedRandomCutForestState state){
```

Comment:
Does it have to be a BasicThresholder, or can we pass an IThresholder here?

Reply:
Fixing.

Reply:
Fixing.